delta-lake/delta-spark341db/src/main/scala/com/databricks/sql/transaction/tahoe/rapids/GpuLowShuffleMergeCommand.scala

```diff
@@ -0,0 +1,1050 @@
+
+        deltaTxn.commit(
+          deltaActions,
+          DeltaOperations.Merge(
+            Option(condition),
+            matchedClauses.map(DeltaOperations.MergePredicate(_)),
+            notMatchedClauses.map(DeltaOperations.MergePredicate(_)),
+            // We do not support notMatchedBySourcePredicates yet and fall back to CPU
+            // See https://github.com/NVIDIA/spark-rapids/issues/8415
+            notMatchedBySourcePredicates = Seq.empty[MergePredicate]
+          ))
+
+        // Record metrics
+        val stats = GpuMergeStats.fromMergeSQLMetrics(
+          metrics, condition, matchedClauses, notMatchedClauses,
+          deltaTxn.metadata.partitionColumns.nonEmpty)
+        recordDeltaEvent(targetDeltaLog, "delta.dml.merge.stats", data = stats)
+
+      }
+      spark.sharedState.cacheManager.recacheByPlan(spark, target)
+    }
+    // This is needed to make the SQL metrics visible in the Spark UI. Also this needs
+    // to be outside the recordMergeOperation because this method will update some metric.
+    val executionId = spark.sparkContext.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)
+    SQLMetrics.postDriverMetricUpdates(spark.sparkContext, executionId, metrics.values.toSeq)
+    Seq(Row(metrics("numTargetRowsUpdated").value + metrics("numTargetRowsDeleted").value +
+      metrics("numTargetRowsInserted").value, metrics("numTargetRowsUpdated").value,
+      metrics("numTargetRowsDeleted").value, metrics("numTargetRowsInserted").value))
+  }
+
+  private def getSourceDF(spark: SparkSession): DataFrame = {
+    // UDF to increment metrics
+    val incrSourceRowCountExpr = makeMetricUpdateUDF("numSourceRows")
+    Dataset.ofRows(spark, source)
+      .filter(new Column(incrSourceRowCountExpr))
+  }
+
+
+  /** Expressions to increment SQL metrics */
+  private def makeMetricUpdateUDF(name: String, deterministic: Boolean = false): Expression = {
+    // only capture the needed metric in a local variable
+    val metric = metrics(name)
+    var u = udf(new GpuDeltaMetricUpdateUDF(metric))
+    if (!deterministic) {
+      u = u.asNondeterministic()
+    }
+    u.apply().expr
+  }
+
+  private def getTargetOutputCols(txn: OptimisticTransaction): Seq[NamedExpression] = {
+    txn.metadata.schema.map { col =>
+      targetOutputAttributesMap
+        .get(col.name)
+        .map { a =>
+          AttributeReference(col.name, col.dataType, col.nullable)(a.exprId)
+        }
+        .getOrElse(Alias(Literal(null), col.name)()
+        )
+    }
+  }
+
+  /**
+   * Repartitions the output DataFrame by the partition columns if table is partitioned
+   * and `merge.repartitionBeforeWrite.enabled` is set to true.
+   */
+  protected def repartitionIfNeeded(
+      spark: SparkSession,
+      df: DataFrame,
+      partitionColumns: Seq[String]): DataFrame = {
+    if (partitionColumns.nonEmpty && spark.conf.get(DeltaSQLConf.MERGE_REPARTITION_BEFORE_WRITE)) {
+      df.repartition(partitionColumns.map(col): _*)
```

Comment:
This logic may need to change to consider optimized write, see #10417

Reply:
That's fine, but probably worth putting a TODO and linking to the issue since we know we're doing it wrong here.